src/TorchSharp/DataLoader.cs

```diff
@@ -112,16 +123,51 @@ public bool MoveNext()
                             DisposeCurrent();
                             if (!MoveNextValue()) return false;
                             List<Dictionary<string, Tensor>> dic = new();
-                            dic.Add(dataset.GetTensor(currentVal));
-                            for (var i = 1; i < batchSize; i++) {
-                                if (!MoveNextValue()) break;
-                                dic.Add(dataset.GetTensor(currentVal));
-                            }
+                            var taskedBatchCount = 0;
+                            var taskBatchLock = new object();
+
+                            TaskBatch();
+                            var t = currentVal;
+
+                            //Run Async
+                            foreach(var _ in Enumerable.Range(1, num_worker - 1))
+                                ThreadPool.QueueUserWorkItem(CreateBatch);
+
+                            dic.Add(dataset.GetTensor(t));
+                            CreateBatch(null);
+
+                            while (dic.Count < taskedBatchCount) { } //Wait for every task finished
 
                             Current = new();
                             foreach (var x in dic[0].Keys)
                                 Current[x] = cat(dic.Select(k => k[x].unsqueeze(0)).ToArray(), 0).to(device);
                             return true;
+
+                            void CreateBatch(object _)
+                            {
+                                while (TaskBatch()) {
+                                    var cv = 0L;
+                                    lock (moveNextLock) {
+                                        if (!MoveNextValue()) {
+                                            break;
```

Comment:
Without building the mental model it is unclear to me if this would work when number of workers does not evenly divide the number of elements in the dataset.